### ISSUE TITLE
fei.py cant load some SER files

### ISSUE THREAD

**Issue by ercius** [hyperspy/hyperspy#1928]
Im using hyperspy 1.3.1 freshly installed from conda today and python 3.5. 

I receive this error when using hs.load() on a recently acquired data set:
```python
~\Anaconda3\envs\hyperspy3\lib\site-packages\hyperspy\io_plugins\fei.py in load_ser_file(filename)
--> 336             data[i] = np.fromfile(f,dtype=np.dtype(data_dtype_list + tag_dtype_list),count=1)
    337             f.seek(offset)
    338         _logger.info("Data info:")

ValueError: setting an array element with a sequence.
```
I did some digging in the source to find out the issue. The issue has to do with the difference between the number of elements in a series file and the __valid__ number of elements. Our TIA software always has a series size of 10 for an EELS acquisition (I have no idea why or how to fix this). So, the `data_offset_array.shape = 10` and has a valid dataset at data_offset_array[0]. Most of the other values in the data_offset_array are actually set to 0. This causes all kinds of issues.

The problem lies in fei.py starting with line 332: https://github.com/hyperspy/hyperspy/blob/79b3bf78b3b5c294feae6f23fa2ddf5fcd4e08fa/hyperspy/io_plugins/fei.py#L332
Here is the full code block:
```python
        for i, offset in enumerate(data_offset_array):
            data[i] = np.fromfile(f,
                                  dtype=np.dtype(
                                      data_dtype_list + tag_dtype_list),
                                  count=1)
            f.seek(offset)
```
This will fail because `f.seek(offset)` when offset == 0 gives the error shown at the beginning.

I added an if statement to check whether the offset is 0 and this fixed the issue.
```python
        for i, offset in enumerate(data_offset_array):
            print('i = {}, offset = {}'.format(i,offset))
            if(offset > 0):
                data[i] = np.fromfile(f,dtype=np.dtype(data_dtype_list + tag_dtype_list),count=1)
            f.seek(offset)
```
Using `header['ValidNumberElements']` and skipping non valid data is another way to do this.

**Comment by thomasaarholt** [comment]
Just in case this explains why TIA is always acquiring a series in EELS:
In TIA, in the PEELS menu (you can find this in the bottom right menu list if its not visible on screen) there's a flap-out triangle. 

Open this and for the various acquisition modes (Acquire, Cumulative etc - can't remember exact names right now) you can choose the number of series and the number of passes.

That said, I've done line profile spectrum imaging in TIA, and that seems to be fine to import into hyperspy.

**Comment by ericpre** [comment]
Thanks for the detailed bug report!

If you are interested in fixing it (it seems that you are half way there already), please make a pull request and it would be easier to discuss the changes on that branch. We have a [developer guide](http://hyperspy.readthedocs.io/en/stable/dev_guide.html) in case you haven't already came across it yet.
Ideally, it would be good to have a very small ser/emi file reproducing the bug and to add it to the tests suite.

**Comment by ercius** [comment]
> Just in case this explains why TIA is always acquiring a series in EELS:
In TIA, in the PEELS menu (you can find this in the bottom right menu list if its not visible on screen) there's a flap-out triangle.
>Open this and for the various acquisition modes (Acquire, Cumulative etc - can't remember exact names right now) you can choose the number of series and the number of passes.
>That said, I've done line profile spectrum imaging in TIA, and that seems to be fine to import into hyperspy.

Thanks for the tip @thomasaarholt . I am aware of how to change this series size in the FEI user interface. Different versions of TIA and the Titan User Interface (a.k.a. ESVision) have different bugs in them. In our version of the UI the Series Size for an Acquire is greyed out and hard coded to 10. It is thus unfixable. See terrible phone screen shot:
![image](https://user-images.githubusercontent.com/15657108/39386705-ae633762-4a2a-11e8-8acb-3097dda55e0a.png)

**Comment by ercius** [comment]
@ericpre 
> Thanks for the detailed bug report!
If you are interested in fixing it (it seems that you are half way there already), please make a pull request and it would be easier to discuss the changes on that branch. We have a developer guide in case you haven't already came across it yet.
Ideally, it would be good to have a very small ser/emi file reproducing the bug and to add it to the tests suite.

I can make a PR for this, and can share a file for testing. It would be useful to discuss whether you want to make a simple fix (like the if statement I added) or utilize the validNumberElements value from the header to only load valid data. This is what I do in my [serReader](https://bitbucket.org/ercius/openncem/src/b3baa098c0af621ddbcc42fced616449bbd8f9ef/serReader.py?at=master&fileviewer=file-view-default#serReader.py-127). 

Ill try to get to this soon.

**Comment by ericpre** [comment]
Great! Indeed using the `ValidNumberElements` from the header sounds like a better fix.
Out of curiosity, I have just tested a stack of images recorded with preview, where the `ValidNumberElements` < `TotalNumberElements` and the file loads fine, but the shape of the loaded data is `x*y*TotalNumberElements` and actually there is no point loading empty data... Using `ValidNumberElements` may be also fixing this as the same time, if used appropriately, maybe?

**Comment by AndrewHerzing** [comment]
I've noticed the same error recently when trying to open some TIA files that only contain images.  Strangely, I don't ever recall seeing this issue in the past so something has either changed with Hyperspy's reader or (more likely) the files that are being saved off of our system.  I'll do a bit more inspection soon to see why this might be occurring.

**Comment by ercius** [comment]
I started working on this earlier this week. The simple fix of using ValidNumberElements in the above mentioned lines did not fully fix this issue. There is a place deeper in the code which also has to change.:
fei.py line 530 change to: `array_shape[i] = header['ValidNumberElements'][0]`
I can create a pull request to fix this and make the other changes needed.

However, the `Signal1D` class still reports 10 spectra at the prompt even when the `sp1.data.shape` is set to 1 using the above fixes. I imagine this will cause issue further down the line in other plotting/processing steps. Any idea where the Signal1D class data size is set?

**Comment by ericpre** [comment]
If I understand correctly the shape of the numpy array is correct but not the one of the axes in the axes_manager? If so, this may be due to the list of axes in the dictionary used to generate the signal and this [line](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_patch/hyperspy/io_plugins/fei.py#L515) in may also need to be changed.

**Comment by ericpre** [comment]
@ercius: just wondering if you think you could get this done for the end of August (target for the next release, see #2005)? Or would you prefer that we do it? If so, could you please share a file to reproduce?

**Comment by ercius** [comment]
@ericpre Sorry for the long delay on this. I was hoping to be able to get this done earlier, but had lots of things including vacation get in the way.

You were mostly right. [This line](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_patch/hyperspy/io_plugins/fei.py#L515) was the issue. I changed it as you suggested and now the data is loaded with the correct size.

However, this breaks 3D hyperspectral data and might also affect 2D line scans. Im working on figuring out a good way to enforce ValidNumberElements without breaking other things. Ill setup a pull request sometime today. 

Do you have a preferred way to share data sets? I can send a few spectra which allow you to recreate this issue.

**Comment by ericpre** [comment]
If this is just for temporary sharing, we don't have any preference, any online service should do.
If you want to add very small files to use in the tests, you can put them in the folder [hyperspy/hyperspy/tests/io/FEI_new](https://github.com/hyperspy/hyperspy/tree/RELEASE_next_minor/hyperspy/tests/io/FEI_new) and they should be used in [hyperspy/hyperspy/tests/io/test_fei.py](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/tests/io/test_fei.py). Actually, it would be good to have tests to cover the issue you are fixing. Usually, writing the test as the same time than fixing the issue is quite useful and efficient.
The tests are failing in travis and appveyor at the moment because of a deprecation warning from the last release of numpy (this is fixed in #2025), so don't worry if CI appear red in your PR!

**Comment by ercius** [comment]
Thanks for the information.

I fixed the issue and have uploaded my code to a fork in my own repo. I use GIT for my own personal code and some public code, but have never done a PR. I will need some help.

1. Which branch do you want the PR to land in?
2. I added some data files as requested. I am unfamiliar with writing tests. How can I test my test? What do you use for running the tests?

You can see the code [here](https://github.com/ercius/hyperspyPAE/blob/fei_validNumberElements/hyperspy/io_plugins/fei.py)

Once I know which branch you want to pull this into Ill create a PR.

**Comment by ericpre** [comment]
Sounds great.

> 1. Which branch do you want the PR to land in?

Since this is bug fix, it should go to `RELEASE_next_patch`. Basically, `RELEASE_next_major` is ahead of `RELEASE_next_minor`, which is ahead of `RELEASE_next_patch`. If you have branch out of `RELEASE_next_minor`, which the default branch, you will need to do rebase, but I can do this later. Rebasing may be not trivial sometime...

> 2. I added some data files as requested. I am unfamiliar with writing tests. How can I test my test? What do you use for running the tests?

We use pytest, for run all the tests, run `pytest --mpl` in the test root folder or `pytest --pyargs hyperspy --mpl`. To run a single test file, you can run `pytest test_fei.py` from the `hyperspy/tests/io` folder, for example. Some IDE also have their own way of running tests.
In this case, it should be enough to load a test file and check a few things such as metadata, shape, calibration, depending on what is sensible.
If you find something that there is something not clear in the [developer guide](http://hyperspy.readthedocs.io/en/stable/dev_guide.html), please feel free to comments and/or change it.

**Comment by ercius** [comment]
@ericpre @magnunor 

Thanks for continuing to bug (pun intended) me about this. I just finished something yesterday, and I started working on this today.

I properly forked hyperspy to [ercius/hyperspy](https://github.com/ercius/hyperspy) today. I reimplemented my changes from my old non-fork which fixes this issue for my test data sets. Im on the RELEASE_nest_minor branch.

Also, my branch properly loads data in [15.15.50.Scanning.Acquire.zip](https://github.com/hyperspy/hyperspy/files/3113027/15.15.50.Scanning.Acquire.zip) from #2133.

Please check my branch and see if it fixes the other issues people are having.

I still need to write and run the unit tests. I have not done this before, but it should not be a problem to learn that. Ill also add some data to the repo which can test this issue. Ill try to acquire some very small files today which break without this patch.

**Note:**
There is still one issue where 3D spectral maps can have the incorrect shape. I have a spectral map which should be 10x10x2048 but hs loads it as 100x100x2048. Im having trouble understanding the code at around line 585 in fei.py. I think there is an issue with the load_only_data() function which gets run for some data and not for others. This could use a clean up to make sure that spectra data and image data are loaded with different code to ensure its loaded correctly. There is also a note in there which references this issue, but only set the data to zeros rather than loading it correctly.

**Comment by ercius** [comment]
I got the tests_fei.py working. I had to remove a couple of lines of code and now everything seems to work. Ill add a test and some data for this specific issue and then it should be good to go pending passing the other tests.

**Comment by magnunor** [comment]
One note about including datasets for the test: it should be as small as possible, as big binary files makes the git repository increase in size pretty quickly.

In my experience, the best way is acquiring datasets with very small amount of probe positions and image size. In addition the dataset(s) should ideally have the same value (for example only zeros), since that compresses really easily. For one of my Medipix3 test datasets, the raw file was 8.6 GB, while the compressed one was around 25 kB.

**Comment by ericpre** [comment]
@ercius: generating smallest possible test files as mentioned by @magnunor is indeed not a luxury. We already have fairly decent coverage for the emi/ser reader so only a few file with the "incomplete" dataset may be necessary. I wanted to acquire some of it while on the microscope but I keep forgetting doing it...

> Note:
There is still one issue where 3D spectral maps can have the incorrect shape. I have a spectral map which should be 10x10x2048 but hs loads it as 100x100x2048. Im having trouble understanding the code at around line 585 in fei.py. I think there is an issue with the load_only_data() function which gets run for some data and not for others. This could use a clean up to make sure that spectra data and image data are loaded with different code to ensure its loaded correctly. There is also a note in there which references this issue, but only set the data to zeros rather than loading it correctly.

In your original branch, I had a few suggestions of changes, so please make a PR and we can discuss it there. The issue you mentioned here may be different from what I have seen.

**Comment by ercius** [comment]
I just created the pull request. All test_fei.py tests pass. Im not sure how it might affect other parts of the code as some tests did not pass. Im not sure if this is due to my changes.

I created small data files for the test. I hope they are small enough. In one case the EMI file is very large so I left it out. I would have to find another data set if you really want the EMI file.

The issues with maps is now fixed as well. I moved the changes to the load_only_data() function which was written as a bandaid to this issue, but did not really fix it. I now crop the data for image/spectra series files since the data can only for from 0:ValidNumberElements. This will save on memory especially for when the Series Size value in TIA is set to some random large number to avoid the "first in first out" effect and the user may lose the first image(s) during acquisition.

@ericpre Please let me know what changes you suggest!

**Comment by ericpre** [comment]
Fixed in #2160.
